### ISSUE TITLE
[REQUEST] Delete read chapters

### ISSUE THREAD

**Issue by monagarcia** [Paperback-iOS/app#287]
**Description**
Toggle to automatically delete downloaded chapters that are read.

**How would it improve the App?**
Right now, we have to manually select read chapters > delete downloaded. Even a manual “delete read downloaded chapters” button could work in addition to the existing “delete downloaded chapters.”

**Additional Context**
Can also have an option to specify how many chapters to last read to delete automatically.

**Comment by ConradWeiser** [comment]
I'd imagine this would be inconvenient for some users who want to keep their content just in case they want to re-read it later. This would imply that there should be some kind of toggle in the settings which automatically removes manga. 

There are some inherent dangers with this.
- What if the user is out of internet range, and accidentally __closes__ a manga that they're reading by mistake. Suddenly their content is gone.
- If the system is locked a kind of 'delete only if the user pulls to the next chapter' - This means that the last chapter will never have a valid trigger to be deleted

Do you have a suggestion to mitigate both of these issues?

**Comment by DishankJ** [comment]
What if there were options like delete the second last or third last chapter when read for example if chapter 1,2 and 3 are downloaded and the user reads chapter 1 & 2, only chapter one will get deleted and hence the user will still have access to the last read chapter.

**Comment by Icnopandaz** [comment]
> I'd imagine this would be inconvenient for some users who want to keep their content just in case they want to re-read it later. This would imply that there should be some kind of toggle in the settings which automatically removes manga.
> 
> There are some inherent dangers with this.
> 
> * What if the user is out of internet range, and accidentally **closes** a manga that they're reading by mistake. Suddenly their content is gone.
> * If the system is locked a kind of 'delete only if the user pulls to the next chapter' - This means that the last chapter will never have a valid trigger to be deleted
> 
> Do you have a suggestion to mitigate both of these issues?
> We're trying not to overdo Paperback. To give users a seamless, **simple** experience. Will this feature simply be added to a clutter of settings which is niche to the point where a majority of users won't use it?
> What do you think?

Thinking of a solution as there are many issues with this idea - I believe a setting that allows it to automatically delete a chapter after X amount of time has surpassed in the case people want that feature, but not something like straight after or a few chapters after, but like 3 days for example. Having it as a setting with a time dropdown. 

E.g. (Not specific to Paperback, specifically an example in general) - In Options under "Downloads"  > "Delete Chapter After Read" > (Has a dropdown) for "Delete after": 12h, 24h, 48h, 72h.) 

I feel that is in a way as simple as it could be after a lot of thought, obviously having it disabled by default but having the setting there for those that want it under the assumption its a highly requested QOL feature, as you say you don't necessarily want heaps of clutter there. (Unless you had an "Advanced" Settings menu but I digress.)

**Comment by monagarcia** [comment]
> I'd imagine this would be inconvenient for some users who want to keep their content just in case they want to re-read it later. This would imply that there should be some kind of toggle in the settings which automatically removes manga.
> 
> There are some inherent dangers with this.
> 
> * What if the user is out of internet range, and accidentally **closes** a manga that they're reading by mistake. Suddenly their content is gone.
> * If the system is locked a kind of 'delete only if the user pulls to the next chapter' - This means that the last chapter will never have a valid trigger to be deleted
> 
> Do you have a suggestion to mitigate both of these issues?
> We're trying not to overdo Paperback. To give users a seamless, **simple** experience. Will this feature simply be added to a clutter of settings which is niche to the point where a majority of users won't use it?
> What do you think?


My comment on the *additional context*, I think, answers this. 
“Can also have an option to specify how many chapters to last read to delete automatically.”

Eg: Delete 5 chapters to last read automatically. If you’re on chapter 15, chapters 1-9 will be deleted but chapters 10-15 will remain. 

I also like the suggestion above to have the option to delete after x hours. 

I think a lot of users will enjoy having this. Coming from a previous Android user who used a well-known manga app there, this is one of the features that people seemed to love to have as a QOL.

**Comment by Icnopandaz** [comment]
> My comment on the _additional context_, I think, answers this.
> “Can also have an option to specify how many chapters to last read to delete automatically.”

That's true, I'm just curious how it'd fit in terms of the settings for the end-user, but also how it'd be coded. E.g. - If I open Chapter 55 by mistake (Lets say at the top of searching the anime it showed the most recent chapter, which iirc it does have when searching for an anime?) would it delete all chapters before 51, or would it specifically remember and delete the last 5 chapters that you've opened? If the later, doing this in such a way that it remembers the last 5 individually per manga, because not being per manga wouldn't work either. I may be overthinking it of course, but am plainly thinking about a solution that isn't just simple for the end-user but simple for the Developer as well. Based on Conrad mentioning a **simple** system I'm thinking of both front and back-ends. 

I could be overthinking it in all honesty, but it sounds like a bit of a development nightmare potentially. - Back to overthinking though: (I'm really playing devils advocate here, Realistically most of these can be solved by having the feature disabled by default, and having it as an advanced setting, but I digress.) 

> What if the user is out of internet range, and accidentally closes a manga that they're reading by mistake. Suddenly their content is gone.

> “Can also have an option to specify how many chapters to last read to delete automatically.”

All it takes is for someone reading their manga to put their device down, have a little kid come over and spam the arrow button and the downloaded manga is gone (Weird example but I digress), I would say you answered the concern but out of the picture there are many other concerns with it, accounting for how people are.

> If the system is locked a kind of 'delete only if the user pulls to the next chapter' - This means that the last chapter will never have a valid trigger to be deleted

Using the idea Monagarcia used- If the system is based on after x chapters (lets say 5) and its checked after each chapter (At the start of the following chapter I'd say in order to minimize the chance of the trigger not happening) then that may work, obviously you don't want it to check and trigger too often. 

> I think a lot of users will enjoy having this. Coming from a previous Android user who used a well-known manga app there, this is one of the features that people seemed to love to have as a QOL.

I still personally use Tachiyomi on my phone (Android), use Paperback on my iPad (Waiting for Webtoons Source from Android to come to Paperback primarily before I fully move over) - Definitely great to take influence and features from existing products for QOL changes :D

**Comment by monagarcia** [comment]
This all very true especially with how Paperback handles “read” chapters, which makes it hard to differentiate what really is *done reading* vs *reading/opened*.

If the main issue is with the feature being automatic, and to avoid accidents, I guess we can just have the option to manually “Delete all read downloaded chapters,” together with the existing “Delete all downloaded chapters.”

**Comment by tejassharma96** [comment]
> What if the user is out of internet range, and accidentally closes a manga that they're reading by mistake. Suddenly their content is gone.
> That's true, I'm just curious how it'd fit in terms of the settings for the end-user, but also how it'd be coded. E.g. - If I open Chapter 55 by mistake (Lets say at the top of searching the manga it showed the most recent chapter, which iirc it does have when searching for a manga?) would it delete all chapters before 51, or would it specifically remember and delete the last 5 chapters that you've opened?  

Regarding how to tell if a chapter is actually completed - paperback already stores which page number you're on (see screenshot). It could just check if you’ve read all the pages and only delete if that’s the case. As another user of the tachiyomi, I loved this feature as it meant I could set it to automatically download new chapters of manga, and then they’d be deleted once I read them.

> All it takes is for someone reading their manga to put their device down, have a little kid come over and spam the arrow button and the downloaded manga is gone (Weird example but I digress), I would say you answered the concern but out of the picture there are many other concerns with it, accounting for how people are.

I don’t believe this is really a valid argument - if someone messes up and marks a bunch of chapters as read and they get deleted as a result, then the feature is working as expected. This definitely shouldn’t be enabled by default, but if someone enables automatic deletion of read chapters and their read chapters get deleted, that’s on them…

> Will this feature simply be added to a clutter of settings which is niche to the point where a majority of users won't use it?

Can’t really speak about the majority of users obviously, and I don’t think tachiyomi has analytics on usage, but I strongly feel that it won’t be clutter - automating something like this feels like a really basic feature that paperback is lacking to me.

> Thinking of a solution as there are many issues with this idea - I believe a setting that allows it to automatically delete a chapter after X amount of time has surpassed in the case people want that feature, but not something like straight after or a few chapters after, but like 3 days for example. Having it as a setting with a time dropdown.

I kinda feel like this is over complicating the issue. But, that might just be because a) I’d never use something like that, b) the devs would need to make a decision on when to check for the time stamps - they would need to check the downloaded chapters for each manga in the library and compare the read timestamps, which could be a slow operation if you have a lot of manga/a lot of chapters downloaded. At least with checking based on number of chapters it’s pretty straightforward - you’d check when the user goes to a new chapter and when they close the reader, and you’d only check the manga currently being read, not every manga in the library. 

![16B99E97-97FE-4201-B625-F5585C94D5FD](https://user-images.githubusercontent.com/8764853/128675809-9e266c63-dd37-4b0b-b61e-731ab912b883.png)

Sorry about the word vomit lol, I just feel strongly about this feature.

**Comment by hyoretsu** [comment]
Bump for awareness, this would really help us users with bad internet on 64GB devices

**Comment by celarye** [comment]
As part of a repository-wide cleanup for the 0.9 release, we are resetting our issue tracker to more efficiently track new feedback and feature requests. If this issue is still relevant or critical to your workflow, please feel free to reopen it with updated context. Thank you for your support!
